### PR TITLE
2432 Add RandCropByLabelClasses transform

### LINKED ISSUE #2432: Support to set pos/neg crop ratio for every class
**Is your feature request related to a problem? Please describe.**
Feature request from @myron .
currently, we can only set the ratio for foreground / background, can't set the ratio for every class.
Need to develop a new crop transform to support ratio list of every class.

### PR DESCRIPTION
**Pull request by Nic-Ma**
Fixes #2432 #2560 .

### Description
This PR added the `RandCropByLabelClasses` transform.

### Status
**Ready**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

File: monai/transforms/utils.py [hunk 1]
```diff
@@ -270,9 +272,66 @@ def map_binary_to_indices(
...
+        for c in range(channels):
+            label_flat = np.any(label[c : c + 1], axis=0).ravel()
+            _add_indices(label_flat=label_flat)
+    else:
+        # Argmax label
+        if num_classes is None:
+            raise ValueError("if not One-Hot format label, must provide the num_classes.")
+        for c in range(num_classes):
+            label_flat = np.any(label == c, axis=0).ravel()
+            _add_indices(label_flat=label_flat)
```
Comment (wyli) [inline monai/transforms/utils.py:330]: I think these two cases could be merged to improve the readability

File: monai/transforms/utils.py [hunk 2]
```diff
@@ -317,6 +376,45 @@ def weighted_patch_samples(
...
+
+    # Select subregion to assure valid roi
+    valid_start = np.floor_divide(spatial_size, 2)
+    # add 1 for random
+    valid_end = np.subtract(label_spatial_shape + np.array(1), spatial_size / np.array(2)).astype(np.uint16)
+    # int generation to have full range on upper side, but subtract unfloored size/2 to prevent rounded range
+    # from being too high
+    for i, valid_s in enumerate(
+        valid_start
+    ):  # need this because np.random.randint does not work with same start and end
```
Comment (wyli) [inline monai/transforms/utils.py:403]: format could be improved?

File: monai/transforms/utils.py [hunk 3]
```diff
@@ -392,7 +463,53 @@ def _correct_centers(
...
+    indices = [np.asarray(i) for i in indices]
+    for i, array in enumerate(indices):
+        if len(array) == 0:
+            warnings.warn(f"no available indices of class {i} to crop, set the crop ratio of this class to zero.")
+            ratios[i] = 0
+
+    centers = []
+    for _ in range(num_samples):
+        # randomly select the indices of a class based on the ratios
+        indices_to_use = indices[rand_state.choice(len(ratios), p=np.asarray(ratios) / np.sum(ratios))]
```
Comment (wyli) [inline monai/transforms/utils.py:507]: the `ratios` needs more input validation.
also I think `rand_state.choice(len(ratios), size= num_samples, p=np.asarray(ratios) / np.sum(ratios))` before the for loop will be slightly more efficient

File: monai/transforms/croppad/dictionary.py [hunk 4]
```diff
@@ -1091,9 +1093,162 @@ def __call__(self, data: Mapping[Hashable, np.ndarray]) -> List[Dict[Hashable, n
...
+    Dictionary-based version :py:class:`monai.transforms.RandCropByLabelClasses`.
+    Crop random fixed sized regions with the center being a class based on the specified ratios of every class.
+    The label data can be One-Hot format array or Argmax data. And will return a list of arrays for all the
+    cropped images. For example, crop two (3 x 3) arrays from (5 x 5) array with `ratios=[1, 2, 3, 1]` can be::
+
+        [[[0, 0, 0, 0, 0],
+          [0, 1, 2, 1, 0],            [[0, 1, 2],     [[0, 0, 0],
+          [0, 1, 3, 0, 0],     -->     [0, 1, 3],      [1, 2, 1],
+          [0, 0, 0, 0, 0],             [0, 0, 0]]      [1, 3, 0]]
+          [0, 0, 0, 0, 0]]]
```
Comment (wyli) [inline monai/transforms/croppad/dictionary.py:1156]: could you put a python runnable/reproducible example here?

**Review by wyli — COMMENTED** [review]
thanks I have some minor comments

File: monai/transforms/utils.py (hunk 1 above)
Comment (Nic-Ma) [reply monai/transforms/utils.py:330]: Sure, updated in the latest commit.
Thanks

File: monai/transforms/utils.py (hunk 2 above)
Comment (Nic-Ma) [reply monai/transforms/utils.py:403]: sure, updated in the latest commit.
Thanks.

File: monai/transforms/utils.py (hunk 3 above)
Comment (Nic-Ma) [reply monai/transforms/utils.py:507]: added non-negative check for ratios.

I think every sample should:
(1) randomly choose which class to select center.
(2) randomly select a center to crop.

So we can't move the `rand_state.choice(len(ratios), size= num_samples, p=np.asarray(ratios) / np.sum(ratios))` before the for loop.

Thanks.

File: monai/transforms/croppad/dictionary.py (hunk 4 above)
Comment (Nic-Ma) [reply monai/transforms/croppad/dictionary.py:1156]: Sure, added a tiny example code in the latest commit.

Thanks

**Comment by Nic-Ma** [comment]
Hi @wyli ,

Thanks for your review.
I updated the PR according to your comments.
Could you please help review it again?

Thanks in advance.

File: monai/transforms/utils.py (hunk 3 above)
Comment (wyli) [reply monai/transforms/utils.py:500]: I think for (1) generating one sample within the loop is equivalent to generating 'num_samples' before the loop and indexing the samples

Comment (Nic-Ma) [reply monai/transforms/utils.py]: Oh, yes, you are right.
I updated in the latest commit.

Thanks.

File: monai/transforms/utils.py [hunk 5]
```diff
@@ -392,7 +453,59 @@ def _correct_centers(
...
-        centers.append(_correct_centers(center_ori, valid_start, valid_end))
+        centers.append(correct_crop_centers(center_ori, spatial_size, label_spatial_shape))
+
+    return centers
+
+
+def generate_label_classes_crop_centers(
+    spatial_size: Union[Sequence[int], int],
+    num_samples: int,
+    ratios: List[Union[float, int]],
```
Comment (wyli) [inline monai/transforms/utils.py:464]: might be useful to set `ratios` to `ones(len(indices))` by default?

Comment (Nic-Ma) [reply monai/transforms/utils.py]: Sure, added default values in the latest commit.

Thanks.

**Review by wyli — APPROVED** [review]

**Pull request merged**